pkg/controllers/jobset_controller.go

```diff
@@ -364,31 +369,23 @@ func (r *JobSetReconciler) restartPolicyRecreateAll(ctx context.Context, js *job
 	return nil
 }
 
-func (r *JobSetReconciler) deleteJobs(ctx context.Context, js *jobset.JobSet, jobsForDeletion []*batchv1.Job) error {
+func (r *JobSetReconciler) deleteJobs(ctx context.Context, jobsForDeletion []*batchv1.Job) error {
 	log := ctrl.LoggerFrom(ctx)
-
-	var wg sync.WaitGroup
+	lock := &sync.Mutex{}
 	var finalErr error
-
-	// Delete all jobs in parallel.
-	// TODO: limit number of goroutines used here.
-	for _, job := range jobsForDeletion {
-		job := job
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-
-			// Delete job. This deletion event will trigger another reconcilliation,
-			// where the jobs are recreated.
-			backgroundPolicy := metav1.DeletePropagationBackground
-			if err := r.Delete(ctx, job, &client.DeleteOptions{PropagationPolicy: &backgroundPolicy}); client.IgnoreNotFound(err) != nil {
-				finalErr = err
-				return
-			}
-			log.V(2).Info("successfully deleted job", "job", klog.KObj(job), "restart attempt", job.Labels[job.Labels[RestartsKey]])
-		}()
-	}
-	wg.Wait()
+	workqueue.ParallelizeUntil(ctx, parallelDeletions, len(jobsForDeletion), func(i int) {
+		targetJob := jobsForDeletion[i]
+		// Delete job. This deletion event will trigger another reconciliation,
+		// where the jobs are recreated.
+		backgroundPolicy := metav1.DeletePropagationBackground
+		if err := r.Delete(ctx, targetJob, &client.DeleteOptions{PropagationPolicy: &backgroundPolicy}); client.IgnoreNotFound(err) != nil {
+			lock.Lock()
+			defer lock.Unlock()
+			finalErr = errors.Join(finalErr, err)
```

Comment:
Oh, `errors.Join()` is introduced since Go 1.20, although this project uses Go 1.19.

Reply:
@ahg-g @danielvegamyhre I'd like to update the Go version to 1.20 in another PR before merging this one into the main. 

Does bumping the Go version to v1.20 sound good?

Reply:
Sounds good to me, let's wait for Abdullah to confirm as well though.

Reply:
Sure :)

Reply:
I will say that I remember upgrading the e2e tests in test infra to 1.20

Reply:
> I will say that I remember upgrading the e2e tests in test infra to 1.20

That makes sense.

Reply:
> @ahg-g @danielvegamyhre I'd like to update the Go version to 1.20 in another PR before merging this one into the main.
> 
> Does bumping the Go version to v1.20 sound good?

@tenzen-y  We can do update it like k/k repos, adding a .goversion file? or we have other ways? just curious

Reply:
@charles-chenzz We can update it by just bumping in go.mod. And then, we need to update it in k/test-infra as mentioned by @kannon92. However, we don't decide to bump it yet.

https://github.com/kubernetes-sigs/jobset/blob/1ef251cb5e72a289e540883559f2c2bc28191178/go.mod#L3

Reply:
I meant that I think we are already using a go1.20 for the e2e tests

Reply:
Thanks!

Reply:
https://github.com/kubernetes-sigs/jobset/pull/129